scripts/prepare_for_registration_bulk_creation.py

```diff
@@ -0,0 +1,117 @@
+    if not initiator:
+        logger.error('Initiator not found: [id={}]'.format(initiator_id))
+        return handle_error(initiator, inform_product=True)
+
+    try:
+        provider = RegistrationProvider.objects.get(_id=provider_id)
+    except RegistrationProvider.DoesNotExist or RegistrationProvider.MultipleObjectsReturned:
+        logger.error('Registration provider not found or multiple results returned: [_id={}] '.format(provider_id))
+        return handle_error(initiator, inform_product=True)
+
+    if not parsing_output:
+        logger.error('Missing task input: `parsing_output`')
+        return handle_error(initiator, inform_product=True)
+
+    schema_id = parsing_output.get('schema_id', None)
+    try:
+        schema = RegistrationSchema.objects.get(_id=schema_id)
+    except RegistrationProvider.DoesNotExist or RegistrationProvider.MultipleObjectsReturned:
+        logger.error('Registration schema not found or multiple results returned: [_id={}] '.format(schema_id))
+        return handle_error(initiator, inform_product=True)
+
+    upload = RegistrationBulkUploadJob.create(payload_hash, initiator, provider, schema)
+    logger.info('Inserting the bulk upload job [hash={}] into table '
+                'RegistrationBulkUploadJob ...'.format(upload.payload_hash))
+    if not dry_run:
+        try:
+            upload.save()
+        except ValidationError as e:
+            logger.error('Insertion failed: [error={}, message={}]'.format(type(e), e.messages))
+            return handle_error(initiator, inform_product=True)
+        upload.reload()
+        logger.info('Insertion successful: [pk={}, hash={}] '.format(upload.id, upload.payload_hash))
+    else:
+        logger.info('Dry run: insertion not saved')
+
+    registration_rows = parsing_output.get('registrations', [])
+    if not registration_rows:
+        logger.error('Missing registration rows')
+        return handle_error(initiator, inform_product=True)
+
+    logger.info('Preparing [{}] registration rows for bulk creation ...'.format(len(registration_rows)))
+    bulk_upload_rows = []
+    for registration_row in registration_rows:
+        bulk_upload_row = RegistrationBulkUploadRow(upload=upload, draft_registration=None, is_completed=False,
+                                                    is_picked_up=False, csv_raw=registration_row.get('csv_raw', ''),
+                                                    csv_parsed=registration_row.get('csv_parsed'))
+        bulk_upload_rows.append(bulk_upload_row)
+
+    if dry_run:
+        logger.info('Dry run: bulk insertion not run')
+        logger.info("Dry run: complete")
+        return
+
+    try:
+        logger.info('Bulk inserting [{}] registration rows into table '
+                    'RegistrationBulkUploadRow ...'.format(len(bulk_upload_rows)))
+        created_objects = RegistrationBulkUploadRow.objects.bulk_create(bulk_upload_rows)
+    except (ValueError, IntegrityError) as e:
```

Comment:
Are `ValueError` and `IntegrityError` the only possible errors? Or do we want do catch all exceptions?

Reply:
Good question. I spent some time on this one. The one draw back of `bulk_create()`is I could not find useful documentation or guide on what exceptions to expect. I looked a few levels deeper and it is still hard to tell. In our code base, I haven't seen any try and except on other `bulk_create`, which I am surprised.

I had the following in my original code but then read about it is not recommended. And we usually just catch the most possible / related exceptions. I would love to add this back if you also think it is ok. This is a limitation of `Python`.

```
except Exception as e:
    ...
```

Reply:
I guess you're right. I was thinking about what happens when there is an uncaught exception. But since uncaught exceptions will always halt program execution, I guess it is fine here to just catch the related ones.